sprint 10 first

alexhved opened the pull request:
sprint 10 homework

alexhved committed:
sprint 10 first

alexhved committed:
sprint 10 fix validate tests

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    @ExceptionHandler(IllegalArgumentException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
```
400 ошибка обрабатывается два раза. Здесь и на строке 19. Разница лишь в перехватываемых ошибка. Можно сделать один перехватчик этой ошибки. Для этого нужно проверить так ли необходим второй класс ошибки?

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/ErrorHandler.java:
```diff
@@ -0,0 +1,46 @@
+
+@RestControllerAdvice
+@Slf4j
+public class ErrorHandler {
```
Может ли возникнуть 404 ошибка в приложении? Ошибка же для этого кода есть в приложении.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,50 +1,69 @@
 public class FilmController {
-    private final FilmValidator validator = new FilmValidator();
-    private final Map<Long, Film> filmMap = new HashMap<>();
+    private final FilmStorage filmStorage;
```
Контроллер не должен обращаться к стораджу. Контроллер обращается к сервису. Сервис обращается к стораджу.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,50 +1,69 @@
     @PostMapping()
-    public Film create(@Valid @RequestBody Film film) throws ValidateException {
+    public Film create(@RequestBody Film film) throws ValidateException {
         validator.validate(film);
```
Валидацию нужно отправить в сервис.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/exception/IncorrectParameterException.java:
```diff
@@ -0,0 +1,14 @@
+public class IncorrectParameterException extends RuntimeException {
+    private final String param;
+
+    public IncorrectParameterException(String param) {
```
Проверьте, где используется эта ошибка.

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/service/user/UserService.java:
```diff
@@ -0,0 +1,106 @@
+            throw new NotFoundException(String.format("User with id %s not found", id));
+        }
+        if (otherUser == null) {
+            throw new NotFoundException(String.format("Пользователь с id %s не найден", friendId));
```
Лучше писать все логи на одном языке.

alexhved committed:
sprint 10 fix exceptions and UserService

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,50 +1,62 @@
-        return ++idGenerator;
+    @GetMapping("/{id}")
+    public Film findById(@PathVariable Long id) throws NotFoundException {
+        return filmService.getFilmStorage().getFilmById(id);
```
здесь нужно вызвать метод у сервиса и все. Не нужно вызывать экземпляр сервиса и вызывать у него метод.

alexhved replied:
ok

sergeymamonov commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -1,50 +1,62 @@
-        return film;
+    public Film create(@RequestBody Film film) throws ValidateException {
+        filmService.getValidator().validate(film);
+        return filmService.getFilmStorage().create(film);
```
Нужно вызвать только метод у сервиса. А уже в этом методе в сервисе будет вызываться валидация.

alexhved replied:
ok

alexhved committed:
refactor: Из FilmController перенесена валидация в FilmService, из UserController перенесена валидация в UserService

alexhved committed:
refactor: из InMemoryFilmStorage в FilmService перенести idGenerator
из InMemoryUserStorage в UserService перенести idGenerator
fix: FilmValidatorTest, UserValidatorTest

alexhved commented:
Исправлены замечания

alexhved commented:
Исправлены замечания описанные выше
